Task5/Task5_2/Catalog.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System.Collections;
+using System.Text.RegularExpressions;
+
+namespace Task5_2
+{
+    internal class Catalog<isbn, book> : IEnumerable<book>
+    {
+        private Dictionary<isbn, book> _catalog;
+
+        public book this[isbn key]
+        {
+            get
+            {
+                key = SimplifyString(key);
+
+                if (_catalog.ContainsKey(key))
+                {
+                    return _catalog[key];
+                }
+                else
+                {
+                    return default(book);
+                }
+            }
+            set
+            {
+                key = SimplifyString(key);
+
+                _catalog[key] = value;
+            }
+        }
+
+        public Catalog()
+        {
+            _catalog = new Dictionary<isbn, book>();
+        }
+
+        public void Add(isbn key, book value)
+        {
+            key = SimplifyString(key);
+
+            _catalog.Add(key, value);
+        }
+
+        isbn SimplifyString(isbn isbn)
+        {
+            string keyCast = isbn.ToString();
+
+            if (keyCast.Contains("-"))
+            {
+                if (!CheckFormat(keyCast))
+                {
+                    throw new ArgumentException("Isbn is not valid");
+                }
+
+                keyCast = keyCast.Replace("-", "");
+            }
+
```

Comment:
I think that this check is not necessary, try to create one Regex for all checks